---
title: "Analysis by gene"
date: "2020-04-13"
author: "Moreno Martinovic"
---

***

```{r}
library(data.table)
library(ggplot2)
library(biomaRt)
library(GenomicRanges)
library(BSgenome.Hsapiens.UCSC.hg38)
library(stringr)
library(magrittr)
```
&nbsp;  

Reading HIV integrations data and control sites:  

```{r}
IS <- fread("is_hg38.txt")[, c(1:4)]
colnames(IS) <- c("chr", "start", "end", "dataset")
IS[, RMC := "Experimental"]
setkey(IS, chr, start, end)
```
&nbsp;  

Getting hg38 genes:  

```{r}
ensembl <- useMart("ensembl")
ensembl <- useDataset("hsapiens_gene_ensembl", mart = ensembl)

hg38_genes <- data.table(getBM(attributes = c("ensembl_gene_id", "chromosome_name", "start_position", "end_position",                                                                               "gene_biotype", "strand"),
                             filters = c("biotype"),
                             values = c("protein_coding", "lncRNA"),
                             mart = ensembl))

setnames(hg38_genes, old = c("start_position", "end_position", "chromosome_name"), new = c("start", "end", "chr"))
hg38_genes[, chr := paste("chr", as.character(chr), sep = "")]
hg38_genes <- hg38_genes[chr %in% paste0("chr", c(1:22, "X"))]
# hg38_genes <- hg38_genes[chr %in% paste0("chr", 1:22) & ((end - start + 1) >= 10000)]

# hg38_genes[, strand := as.character(strand)]
# hg38_genes[strand == "1", strand := "+"]
# hg38_genes[strand == "-1", strand := "-"]
# hg38_genes[, strand := as.factor(strand)]

setkey(hg38_genes, chr, start, end)
```
&nbsp;  

Tiling protein coding genes into 100 bp tiles:  

```{r}
# chrlens <- seqlengths(seqinfo(BSgenome.Hsapiens.UCSC.hg38))
# chroms <- paste0("chr", c(as.character(1:22), "X", "Y"))
# chrlens <- chrlens[names(chrlens) %in% chroms]
# 
# tiles100bp <- as.data.table(unlist(tileGenome(chrlens - (chrlens %% 100), tilewidth = 100)))[, .(seqnames, start, end)]
# setnames(tiles100bp, "seqnames", "chr")
# setkey(tiles100bp, chr, start, end)

# ovlps <- foverlaps(tiles100bp, prot_coding)
# prot_coding_100bp <- ovlps[!is.na(start),
#                            .(chr, start = i.start, end = i.end, ensembl_gene_id, hgnc_symbol, integrationsInGene = numberOfIntegrations)]
```
&nbsp;  

Making regions -2000 and +2000 from TSS and splitting genes by %:  

```{r}
hg38_genes <- hg38_genes[, .(chr, start, end, strand, ensembl_gene_id)]

genes_by_perc <- 
  hg38_genes[,
             {
               flankleft_starts <- as.integer(seq(start - 2000, start, length.out = 34))
               flankleft_ends <- flankleft_starts[-34] + diff(flankleft_starts) - 1
               flankleft_starts <- flankleft_starts[-34]
               
               starts <- as.integer(seq(start, end + 1, length.out = 101))
               ends <- starts[-101] + diff(starts) - 1
               starts <- starts[-101]
               
               flankright_starts <- as.integer(seq(end + 1, end + 2001, length.out = 34))
               flankright_ends <- flankright_starts[-34] + diff(flankright_starts) - 1
               flankright_starts <- flankright_starts[-34]
               
               .(start = c(flankleft_starts, starts, flankright_starts),
                 end = c(flankleft_ends, ends, flankright_ends))
             },
             by = .(chr, gene_start = start, gene_end = end, strand, ensembl_gene_id)]

genes_by_perc <- genes_by_perc[order(as.integer(str_extract(chr, "([0-9])+")), start)]
genes_by_perc[, end := as.integer(end)]

# fwrite(genes_by_perc[, .(chr, start, end)], "genes_by_perc_regions.bed", sep = "\t", col.names = FALSE)
```
&nbsp;  

H3K4me3 coverage by percent gene and flanking regions:   

```{r}
H3K4me3_by_perc <- fread("H3K4me3_profile_TSS/H3K4me3_byperc.bed")[, c(1:3, 5)]
colnames(H3K4me3_by_perc) <- c("chr", "start", "end", "meanCov")

input_by_perc <- fread("H3K4me3_profile_TSS/Input1_byperc.bed")[, c(1:3, 5)]
colnames(input_by_perc) <- c("chr", "start", "end", "meanCov")

H3K4me3_by_perc[, meanCov := meanCov/input_by_perc[, meanCov + 0.001]]

H3K4me3_by_perc <- merge(H3K4me3_by_perc,
                         genes_by_perc[, .(chr, start, end, ensembl_gene_id, strand)],
                         by = c("chr", "start", "end"))

H3K4me3_by_perc <- H3K4me3_by_perc[ensembl_gene_id %in% H3K4me3_by_perc[, .N, ensembl_gene_id][N == 166, ensembl_gene_id]][order(chr, start, end)]

H3K4me3_by_perc[strand == 1, percID := rep(1:166, sum(strand == 1) / 166), by = ensembl_gene_id]
H3K4me3_by_perc[strand == -1, percID := rep(166:1, sum(strand == -1) / 166), by = ensembl_gene_id]

H3K4me3_by_perc[ensembl_gene_id %in% genes_IS[intGroup == "RIGs", ensembl_gene_id], intGrp := "RIG-ovi"]
H3K4me3_by_perc[ensembl_gene_id %in% genes_IS[intGroup == "Not integrated", ensembl_gene_id], intGrp := "Bez integracija"]

# H3K4me3_by_perc[, integrationDensity := numberOfIntegrations / (gene_end - gene_start + 1)]
# H3K4me3_by_perc[, quantile(integrationDensity, 0.95)]
# H3K4me3_by_perc[integrationDensity >= 3.536193e-05, densGroup := "top 5%"]
# H3K4me3_by_perc[integrationDensity > 0 & integrationDensity < 3.536193e-05, densGroup := "0 - top 5%"]
# H3K4me3_by_perc[integrationDensity == 0, densGroup := "0 integrations"]

H3K4me3_meanPerc <- H3K4me3_by_perc[,
                                    {
                                      conf <- as.numeric(t.test(meanCov, conf.level = 0.95)$conf.int)
                                      .(meanCov = mean(meanCov), confLow = conf[1], confHigh = conf[2])
                                    },
                                    by = .(percID, intGrp)]

ggplot(H3K4me3_meanPerc[!is.na(intGrp)], aes(x = percID)) +
  geom_line(aes(y = meanCov, color = intGrp)) +
  geom_ribbon(aes(ymin = confLow, ymax = confHigh, fill = intGrp), alpha = 0.3) +
  scale_x_continuous(breaks = c(0, 33, 66, 99, 132, 165), labels = c("-2000", "TSS", "33%", "66%", "TES", "2000")) +
  labs(x = "5' -> 3'", y = "Prosjeƒçna pokrivenost", color = "", fill = "", title = "H3K4me3") +
  theme_light()

ggsave("H3K4me3_profile.jpg", height = 4.5, width = 8)
```
&nbsp;  

Splitting 10k bp +- from TSS into 100 bp windows:  

```{r}
TSS10k <- 
  hg38_genes[,
             {
               left_starts <- as.integer(seq(start - 10000, start, length.out = 101))[-101]
               left_ends <- left_starts + 99
  
               right_starts <- as.integer(seq(start + 1, start + 10001, length.out = 101))[-101]
               right_ends <- right_starts + 99
               
               .(start = c(left_starts, right_starts),
                 end = c(left_ends, right_ends))
             },
             by = .(chr, gene_start = start, gene_end = end, ensembl_gene_id, gene_biotype, strand)]

TSS10k[, end := as.integer(end)]
```
&nbsp;  

Plotting H3K4me3 +- 10k from TSS:  

```{r}
H3K4me3_TSS10k <- fread("H3K4me3_profile_TSS/H3K4me3_TSS10k.bed")[, c(1:3, 5)]
colnames(H3K4me3_TSS10k) <- c("chr", "start", "end", "meanCov")

input_TSS10k <- fread("H3K4me3_profile_TSS/Input1_TSS10k.bed")[, c(1:3, 5)]
colnames(input_TSS10k) <- c("chr", "start", "end", "meanCov")

H3K4me3_TSS10k[, meanCov := meanCov/input_TSS10k[, meanCov + 0.001]]
H3K4me3_TSS10k <- merge(H3K4me3_TSS10k,
                        TSS10k[, .(chr, start, end, ensembl_gene_id, strand)],
                        by = c("chr", "start", "end"))

H3K4me3_TSS10k <- H3K4me3_TSS10k[ensembl_gene_id %in% H3K4me3_TSS10k[, .N, ensembl_gene_id][N == 200, ensembl_gene_id]][order(chr, start, end)]

H3K4me3_TSS10k[strand == 1, tileID := rep(1:200, sum(strand == 1) / 200), by = ensembl_gene_id]
H3K4me3_TSS10k[strand == -1, tileID := rep(200:1, sum(strand == -1) / 200), by = ensembl_gene_id]

H3K4me3_TSS10k[ensembl_gene_id %in% genes_IS[intGroup == "RIGs", ensembl_gene_id], intGrp := "RIG-ovi"]
H3K4me3_TSS10k[ensembl_gene_id %in% genes_IS[intGroup == "Not integrated", ensembl_gene_id], intGrp := "Bez integracija"]

# H3K4me3_by_perc[, integrationDensity := numberOfIntegrations / (gene_end - gene_start + 1)]
# H3K4me3_by_perc[, quantile(integrationDensity, 0.95)]
# H3K4me3_by_perc[integrationDensity >= 3.536193e-05, densGroup := "top 5%"]
# H3K4me3_by_perc[integrationDensity > 0 & integrationDensity < 3.536193e-05, densGroup := "0 - top 5%"]
# H3K4me3_by_perc[integrationDensity == 0, densGroup := "0 integrations"]

H3K4me3_mean10k <- H3K4me3_TSS10k[,
                                  {
                                    conf <- as.numeric(t.test(meanCov, conf.level = 0.95)$conf.int)
                                    .(meanCov = mean(meanCov), confLow = conf[1], confHigh = conf[2])
                                  },
                                  by = .(tileID, intGrp)]

ggplot(H3K4me3_mean10k[!is.na(intGrp)], aes(x = tileID)) +
  geom_line(aes(y = meanCov, color = intGrp)) +
  geom_ribbon(aes(ymin = confLow, ymax = confHigh, fill = intGrp), alpha = 0.3) +
  scale_x_continuous(breaks = c(0, 99.5, 199), labels = c("-10kb", "TSS", "+10kb")) +
  labs(x = "5' -> 3'", y = "Prosjecna pokrivenost", color = "", fill = "", title = "H3K4me3") +
  theme_light()

ggsave("H3K4me3_TSS_10k_lists.jpg", height = 4.5, width = 8)
```

